src/librustc_data_structures/sync.rs

```diff
@@ -241,6 +242,75 @@ impl<K: Eq + Hash, V: Eq, S: BuildHasher> HashMapExt<K, V> for HashMap<K, V, S>
     }
 }
 
+/// A type whose inner value can be written once and then will stay read-only
+pub struct Once<T>(Lock<Option<T>>, PhantomData<T>);
+
+impl<T> Once<T> {
+    #[inline(always)]
+    pub fn new() -> Self {
+        Once(Lock::new(None), PhantomData)
+    }
+
+    #[inline(always)]
+    pub fn into_inner(self) -> Option<T> {
+        self.0.into_inner()
+    }
+
+    #[inline]
+    pub fn try_set(&self, value: T) -> Option<T> {
+        let mut lock = self.0.lock();
+        if lock.is_some() {
+            return Some(value);
+        }
+        *lock = Some(value);
+        None
+    }
+
+    #[inline]
+    pub fn set(&self, value: T) {
+        assert!(self.try_set(value).is_none());
+    }
+
+    #[inline]
+    pub fn init_locking<F: FnOnce() -> T>(&self, f: F) {
```

Comment:
`init_locking` and `init_nonlocking` should be documented to. I needed a while to see what these are about but the difference can be explained rather quickly.